### PR TITLE
feat: make workspace search bar remember text

### LINKED ISSUE #9474: Remove default owner:me filter for Owners or Template Admin
Hi Coders,

For an admin user of Coder, its kinda annoying removing everytime the filter owner:me from the filter bar when I click the Workspace Tab. AS an admin user, I barely use Coder as an end user, so I'm not interested on my WS, I'm focus on all WS created, running, stopped and abandoned to be removed. When I try to remove 15 abandoned WS, I had to go 1 by 1, clicking on the row, writing the name of the WS (totally unnecessary/annoying), clicking on delete, and going back to WS tab, remove the owner:me filter and repeat 15 times.

image

### PR DESCRIPTION
**Pull request by buenos-nachos**
Closes #9474 

This updates the search bar for the workspace, to ensure that it's able to remember the text the user last used, even after the user closes the tab or goes to a different page.

## File changes made
- Added a polyfill for the forthcoming `useEffectEvent` hook
  - Used it to update the top-level `WorkspacesPage` component to force `setSearchParams` from React Router to be a stable function across renders
- Updated `useFilter` to use a "fallback filter" instead of a single initial value, and have it keep the page's search params in sync as the query changes
  - Updated `useWorkspacesFilter` to try reading the fallback filter from `localStorage`

### CODE REVIEW

**Comment by Parkreiner** [comment]
Basically done with the code changes. Need to do two more things before turning this into a full, proper PR:
- Verify that my change to `setSearchParams` doesn't break anything else that's dependent on it for the `WorkspacesPage` component (99% chance it doesn't, but better to be safe than sorry)
- Write a quick test for the `useEffectEvent` polyfill

**Comment by Parkreiner** [comment]
Going to be checking the logs for the GH Action to see which E2E test is failing

File: site/src/components/Filter/filter.tsx
Comment (BrunoQuaresma) [inline site/src/components/Filter/filter.tsx:60]: Interesting, I thought `setSearchParams` would be stable between render 🤔

File: site/src/components/Filter/filter.tsx
Comment (BrunoQuaresma) [inline site/src/components/Filter/filter.tsx:70]: `useFilterParamsKey` is not very clear to me what it is

File: site/src/components/Filter/filter.tsx
Comment (BrunoQuaresma) [reply site/src/components/Filter/filter.tsx:60]: could `useEffectEvent` be used here?

File: site/src/pages/WorkspacesPage/WorkspacesPage.tsx
Comment (BrunoQuaresma) [inline site/src/pages/WorkspacesPage/WorkspacesPage.tsx:190]: I think `storage` is a better name. I thought `fallbackFilter` was the filter value used in case of failure.

Comment (BrunoQuaresma) [reply site/src/pages/WorkspacesPage/WorkspacesPage.tsx:190]: After re-reading the code I'm confusing what is a fallbackFilter and localStorageFilter 🤔

**Comment by BrunoQuaresma** [comment]
I can understand what is going on but I would like a second eye from @aslilac as well since this is a more complex PR

File: site/src/components/Filter/filter.tsx
Comment (Parkreiner) [reply site/src/components/Filter/filter.tsx:60]: I would've thought so, too, but I had some issues with it making my `useEffect` calls run too often, and I also saw other people complaining about it on StackOverflow

File: site/src/components/Filter/filter.tsx
Comment (Parkreiner) [reply site/src/components/Filter/filter.tsx:70]: Maybe I could get rid of the `use` prefix in the name? Or maybe `urlParamsKey`?

The main goal with the variable is to make sure the same key is being used for all of `currentParams`'s reads/writes (it's already being used in four spots, so definitely don't want typos in the key name breaking things)

File: site/src/pages/WorkspacesPage/WorkspacesPage.tsx
Comment (Parkreiner) [reply site/src/pages/WorkspacesPage/WorkspacesPage.tsx:27]: Actually, I haven't looked. I can go digging through their issues to see if anyone's brought in up in their repo

**Review by BrunoQuaresma — APPROVED** [review]
Approving it to not block you but would be nice to have a second review but up to you. If you feel confident this is the best, go for it and we can discuss it later if it is necessary. Good job!

**Comment by Parkreiner** [comment]
> Approving it to not block you but would be nice to have a second review but up to you. If you feel confident this is the best, go for it and we can discuss it later if it is necessary. Good job!

Yeah, I'm okay with either approach. What do you think, @aslilac ? Not sure what you have on your plate right now

**Pull request merged**
